src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs

```diff
@@ -343,6 +343,27 @@ private SelectItem GenerateExpandItem(ExpandTermToken tokenIn)
             PathSegmentToken currentToken = tokenIn.PathToNavigationProp;
 
             IEdmStructuredType currentLevelEntityType = this.EdmType;
+
+            // If we have a previous level binding with the last segment as a Type segment,
+            // The currentLevelEntityType should be the Type of the TypeSegment
+            // E.g /Orders?$expand=Customer/NS.VipCustomer($expand=Trips)
+            // If we are Binding the expanded property Trips, currentLevelEntityType should be the type of the VipCustomer
+            if (this.parsedSegments.Count > 0 && this.parsedSegments.Last() is TypeSegment)
+            {
+                IEdmType typeSegmentType = this.parsedSegments.Last().EdmType;

```

Comment:
Whatever you decide for the above comment, it would be best to store the result of the last element in a local variable so that we don't have to deference or cast twice

Reply:
#Resolved